File: src/main/services/image.ts
/**
 * 图片管理服务
 * 负责图片的上传、压缩、存储等
 */

import { join, extname } from 'path'
import { existsSync, mkdirSync, unlinkSync, readFileSync } from 'fs'
import { v4 as uuidv4 } from 'uuid'
import sharp from 'sharp'
import { dialog } from 'electron'
import {
  PROJECT_META_DIR,
  IMAGES_DIR
} from '../types/project'
import { ImageFieldConfig } from '../types/vocabulary'

/**
 * 默认图片配置
 */
const DEFAULT_IMAGE_CONFIG: Required<ImageFieldConfig> = {
  maxSize: 5 * 1024 * 1024, // 5MB
  allowedFormats: ['jpg', 'jpeg', 'png', 'gif', 'webp'],
  maxWidth: 1920,
  maxHeight: 1080,
  quality: 85
}

/**
 * 上传结果
 */
export interface ImageUploadResult {
  /** 存储路径（相对于项目根目录） */
  path: string
  /** 原始文件名 */
  originalName: string
  /** 文件大小（字节） */
  size: number
  /** 图片宽度 */
  width: number
  /** 图片高度 */
  height: number
  /** 格式 */
  format: string
}

/**
 * 图片服务
 */
class ImageService {
  /**
   * 获取图片存储目录路径
   */
  private getImagesDir(projectPath: string): string {
    return join(projectPath, PROJECT_META_DIR, IMAGES_DIR)
  }

  /**
   * 确保图片目录存在
   */
  private ensureImagesDir(projectPath: string): string {
    const imagesDir = this.getImagesDir(projectPath)
    if (!existsSync(imagesDir)) {
      mkdirSync(imagesDir, { recursive: true })
    }
    return imagesDir
  }

  /**
   * 从 Base64 上传图片
   * @param projectPath 项目路径
   * @param base64Data Base64 数据（可含 data:image/xxx;base64, 前缀）
   * @param config 图片配置
   */
  async uploadFromBase64(
    projectPath: string,
    base64Data: string,
    config?: ImageFieldConfig
  ): Promise<ImageUploadResult> {
    const finalConfig = { ...DEFAULT_IMAGE_CONFIG, ...config }

    // 解析 Base64 数据
    let data: string
    let format: string

    if (base64Data.startsWith('data:')) {
      // 解析 data URL
      const match = base64Data.match(/^data:image\/(\w+);base64,(.+)$/)
      if (!match) {
        throw new Error('无效的图片数据格式')
      }
      format = match[1].toLowerCase()
      data = match[2]
    } else {
      // 纯 Base64，默认 PNG
      data = base64Data
      format = 'png'
    }

    // 检查格式
    if (!finalConfig.allowedFormats.includes(format)) {
      throw new Error(`不支持的图片格式: ${format}。允许的格式: ${finalConfig.allowedFormats.join(', ')}`)
    }

    // 解码 Base64
    const buffer = Buffer.from(data, 'base64')

    // 检查大小
    if (buffer.length > finalConfig.maxSize) {
      throw new Error(`图片大小超出限制: ${(buffer.length / 1024 / 1024).toFixed(2)}MB > ${(finalConfig.maxSize / 1024 / 1024).toFixed(2)}MB`)
    }

    // 处理图片（压缩、调整尺寸）
    return await this.processAndSave(projectPath, buffer, format, finalConfig)
  }

  /**
   * 从文件路径上传图片
   * @param projectPath 项目路径
   * @param filePath 文件路径
   * @param config 图片配置
   */
  async uploadFromFile(
    projectPath: string,
    filePath: string,
    config?: ImageFieldConfig
  ): Promise<ImageUploadResult> {
    const finalConfig = { ...DEFAULT_IMAGE_CONFIG, ...config }

    // 获取扩展名
    const ext = extname(filePath).toLowerCase().slice(1)
    const format = ext === 'jpg' ? 'jpeg' : ext

    // 检查格式
    if (!finalConfig.allowedFormats.includes(format) && !finalConfig.allowedFormats.includes(ext)) {
      throw new Error(`不支持的图片格式: ${ext}。允许的格式: ${finalConfig.allowedFormats.join(', ')}`)
    }

    // 读取文件
    const buffer = readFileSync(filePath)

    // 检查大小
    if (buffer.length > finalConfig.maxSize) {
      throw new Error(`图片大小超出限制: ${(buffer.length / 1024 / 1024).toFixed(2)}MB > ${(finalConfig.maxSize / 1024 / 1024).toFixed(2)}MB`)
    }

    return await this.processAndSave(projectPath, buffer, format, finalConfig, filePath)
  }

  /**
   * 处理并保存图片
   */
  private async processAndSave(
    projectPath: string,
    buffer: Buffer,
    format: string,
    config: Required<ImageFieldConfig>,
    originalPath?: string
  ): Promise<ImageUploadResult> {
    // 确保目录存在
    const imagesDir = this.ensureImagesDir(projectPath)

    // 使用 sharp 处理图片
    let image = sharp(buffer)
    const metadata = await image.metadata()

    // 计算是否需要缩放
    let needsResize = false
    let width = metadata.width || 0
    let height = metadata.height || 0

    if (width > config.maxWidth || height > config.maxHeight) {
      needsResize = true
    }

    // 缩放图片
    if (needsResize) {
      image = image.resize(config.maxWidth, config.maxHeight, {
        fit: 'inside',
        withoutEnlargement: true
      })
    }

    // 根据格式设置质量
    const outputFormat = format === 'gif' ? 'png' : format as 'jpeg' | 'png' | 'webp'
    const outputExt = format === 'gif' ? 'png' : format

    if (outputFormat === 'jpeg') {
      image = image.jpeg({ quality: config.quality })
    } else if (outputFormat === 'webp') {
      image = image.webp({ quality: config.quality })
    } else {
      image = image.png({ compressionLevel: 6 })
    }

    // 生成文件名
    const filename = `${uuidv4()}.${outputExt}`
    const fullPath = join(imagesDir, filename)

    // 保存文件
    await image.toFile(fullPath)

    // 获取处理后的图片信息
    const processedMetadata = await sharp(fullPath).metadata()
    const stats = await sharp(fullPath).stats()

    return {
      path: `${PROJECT_META_DIR}/${IMAGES_DIR}/${filename}`,
      originalName: originalPath ? originalPath.split(/[/\\]/).pop() || filename : filename,
      size: stats.size || 0,
      width: processedMetadata.width || 0,
      height: processedMetadata.height || 0,
      format: outputExt
    }
  }

  /**
   * 删除图片
   * @param projectPath 项目路径
   * @param imagePath 图片路径（相对于项目根目录）
   */
  async deleteImage(projectPath: string, imagePath: string): Promise<void> {
    const fullPath = join(projectPath, imagePath)
    if (existsSync(fullPath)) {
      unlinkSync(fullPath)
    }
  }

  /**
   * 获取图片的完整路径
   * @param projectPath 项目路径
   * @param imagePath 图片路径（相对于项目根目录）
   */
  getFullImagePath(projectPath: string, imagePath: string): string {
    // 如果是 Base64 数据，直接返回
    if (imagePath.startsWith('data:')) {
      return imagePath
    }
    return join(projectPath, imagePath)
  }

  /**
   * 检查图片是否存在
   * @param projectPath 项目路径
   * @param imagePath 图片路径（相对于项目根目录）
   */
  imageExists(projectPath: string, imagePath: string): boolean {
    // Base64 数据总是存在
    if (imagePath.startsWith('data:')) {
      return true
    }
    return existsSync(join(projectPath, imagePath))
  }

  /**
   * 显示选择图片对话框
   */
  async showOpenDialog(): Promise<string | null> {
    const result = await dialog.showOpenDialog({
      title: '选择图片',
      properties: ['openFile'],
      filters: [
        { name: '图片文件', extensions: ['jpg', 'jpeg', 'png', 'gif', 'webp'] }
      ]
    })

    if (result.canceled || result.filePaths.length === 0) {
      return null
    }

    return result.filePaths[0]
  }

  /**
   * 读取图片为 Base64
   * @param projectPath 项目路径
   * @param imagePath 图片路径（相对于项目根目录）
   */
  async readAsBase64(projectPath: string, imagePath: string): Promise<string> {
    // 如果已经是 Base64，直接返回
    if (imagePath.startsWith('data:')) {
      return imagePath
    }

    const fullPath = join(projectPath, imagePath)
    if (!existsSync(fullPath)) {
      throw new Error('图片文件不存在')
    }

    const buffer = readFileSync(fullPath)
    const ext = extname(imagePath).toLowerCase().slice(1)
    const mimeType = ext === 'jpg' ? 'jpeg' : ext

    return `data:image/${mimeType};base64,${buffer.toString('base64')}`
  }
}

// 导出单例
export const imageService = new ImageService()

File: src-tauri/src/services/image.rs
use crate::error::{AppError, AppResult};
use crate::services::project_state;
use crate::utils::{ensure_dir, generate_id};
use base64::Engine;
use std::io::Cursor;
use std::path::PathBuf;
use std::{fs, io::Write};

const MAX_WIDTH: u32 = 1920;
const MAX_HEIGHT: u32 = 1080;
const JPEG_QUALITY: u8 = 85;

pub struct ImageService;

impl ImageService {
    pub fn new() -> Self {
        Self
    }

    fn get_project_path() -> AppResult<String> {
        project_state::get_project_path().ok_or(AppError::ProjectNotOpen)
    }

    fn get_images_dir(project_path: &str) -> PathBuf {
        project_state::get_data_dir(project_path).join("images")
    }

    fn process_and_save(
        image_data: &[u8],
        format_str: &str,
        images_dir: &std::path::Path,
    ) -> AppResult<serde_json::Value> {
        let img = image::load_from_memory(image_data)
            .map_err(|e| AppError::OperationFailed(format!("解析图片失败: {}", e)))?;

        let original_width = img.width();
        let original_height = img.height();

        let needs_resize = original_width > MAX_WIDTH || original_height > MAX_HEIGHT;
        let img = if needs_resize {
            img.resize(MAX_WIDTH, MAX_HEIGHT, image::imageops::FilterType::Lanczos3)
        } else {
            img
        };

        let format_lower = format_str.to_lowercase();
        let output_format = match format_lower.as_str() {
            "jpg" | "jpeg" => "jpeg",
            "gif" => "png",
            other => other,
        };

        let output_ext = match output_format {
            "jpeg" => "jpg",
            other => other,
        };

        let id = generate_id();
        let filename = format!("{}.{}", id, output_ext);
        let file_path = images_dir.join(&filename);

        let mut encoded_buf = Cursor::new(Vec::new());
        match output_format {
            "jpeg" => {
                let encoder = image::codecs::jpeg::JpegEncoder::new_with_quality(
                    &mut encoded_buf,
                    JPEG_QUALITY,
                );
                img.write_with_encoder(encoder)
                    .map_err(|e| AppError::OperationFailed(format!("编码 JPEG 失败: {}", e)))?;
            }
            "webp" => {
                let encoder = image::codecs::webp::WebPEncoder::new_lossless(&mut encoded_buf);
                img.write_with_encoder(encoder)
                    .map_err(|e| AppError::OperationFailed(format!("编码 WebP 失败: {}", e)))?;
            }
            _ => {
                let encoder = image::codecs::png::PngEncoder::new_with_quality(
                    &mut encoded_buf,
                    image::codecs::png::CompressionType::Default,
                    image::codecs::png::FilterType::Adaptive,
                );
                img.write_with_encoder(encoder)
                    .map_err(|e| AppError::OperationFailed(format!("编码 PNG 失败: {}", e)))?;
            }
        }

        let encoded_data = encoded_buf.into_inner();
        let mut file = fs::File::create(&file_path)
            .map_err(|e| AppError::OperationFailed(format!("创建文件失败: {}", e)))?;
        file.write_all(&encoded_data)
            .map_err(|e| AppError::OperationFailed(format!("写入文件失败: {}", e)))?;

        let relative_path = format!(".novelhelper/data/images/{}", filename);

        Ok(serde_json::json!({
            "path": relative_path,
            "originalName": filename,
            "size": encoded_data.len(),
            "width": img.width(),
            "height": img.height(),
            "format": output_ext
        }))
    }

    fn detect_format_from_data(data: &[u8]) -> String {
        if data.len() < 4 {
            return "png".to_string();
        }
        let header = &data[0..4.min(data.len())];
        if header.starts_with(b"\x89PNG") {
            "png".to_string()
        } else if header.starts_with(b"\xFF\xD8\xFF") {
            "jpeg".to_string()
        } else if header.starts_with(b"GIF8") {
            "gif".to_string()
        } else if header.starts_with(b"RIFF") && data.len() > 11 && &data[8..12] == b"WEBP" {
            "webp".to_string()
        } else {
            "png".to_string()
        }
    }

    pub fn upload_from_base64(
        &self,
        base64_data: String,
        original_name: String,
    ) -> AppResult<serde_json::Value> {
        let project_path = Self::get_project_path()?;
        let images_dir = Self::get_images_dir(&project_path);
        ensure_dir(&images_dir)?;

        let (image_data, format) = if base64_data.starts_with("data:image/") {
            let parts: Vec<&str> = base64_data.splitn(2, ";base64,").collect();
            if parts.len() != 2 {
                return Err(AppError::InvalidParam("无效的 data URL 格式".to_string()));
            }
            let mime_format = parts[0].trim_start_matches("data:image/");
            let decoded = base64::engine::general_purpose::STANDARD
                .decode(parts[1])
                .map_err(|e| AppError::OperationFailed(format!("Base64 解码失败: {}", e)))?;
            (decoded, mime_format.to_string())
        } else {
            let decoded = base64::engine::general_purpose::STANDARD
                .decode(&base64_data)
                .map_err(|e| AppError::OperationFailed(format!("Base64 解码失败: {}", e)))?;
            let ext = PathBuf::from(&original_name)
                .extension()
                .map(|e| e.to_string_lossy().to_string())
                .unwrap_or_else(|| "png".to_string());
            (decoded, ext)
        };

        Self::process_and_save(&image_data, &format, &images_dir)
    }

    pub fn upload_from_file(
        &self,
        file_path: String,
        original_name: String,
    ) -> AppResult<serde_json::Value> {
        let project_path = Self::get_project_path()?;
        let images_dir = Self::get_images_dir(&project_path);
        ensure_dir(&images_dir)?;

        let data = fs::read(&file_path)
            .map_err(|e| AppError::OperationFailed(format!("读取文件失败: {}", e)))?;

        let path = PathBuf::from(&original_name);
        let ext = path
            .extension()
            .map(|e| e.to_string_lossy().to_string())
            .unwrap_or_else(|| Self::detect_format_from_data(&data));

        Self::process_and_save(&data, &ext, &images_dir)
    }

    pub fn delete_image(&self, image_path: String) -> AppResult<()> {
        let project_path = Self::get_project_path()?;
        let full_path = PathBuf::from(&project_path).join(&image_path);
        if full_path.exists() {
            fs::remove_file(&full_path)
                .map_err(|e| AppError::OperationFailed(format!("删除图片失败: {}", e)))?;
        }
        Ok(())
    }

    pub fn read_as_base64(&self, image_path: String) -> AppResult<String> {
        let project_path = Self::get_project_path()?;
        let full_path = PathBuf::from(&project_path).join(&image_path);
        if !full_path.exists() {
            return Err(AppError::FileNotFound(image_path));
        }
        let data = fs::read(&full_path)
            .map_err(|e| AppError::OperationFailed(format!("读取图片失败: {}", e)))?;
        let encoded = base64::engine::general_purpose::STANDARD.encode(&data);

        let ext = PathBuf::from(&image_path)
            .extension()
            .map(|e| e.to_string_lossy().to_lowercase())
            .unwrap_or_else(|| "png".to_string());

        let mime = match ext.as_str() {
            "jpg" | "jpeg" => "image/jpeg",
            "gif" => "image/gif",
            "webp" => "image/webp",
            "svg" => "image/svg+xml",
            "bmp" => "image/bmp",
            _ => "image/png",
        };

        Ok(format!("data:{};base64,{}", mime, encoded))
    }

    pub fn image_exists(&self, image_path: String) -> AppResult<bool> {
        let project_path = Self::get_project_path()?;
        let full_path = PathBuf::from(&project_path).join(&image_path);
        Ok(full_path.exists())
    }

    pub fn get_full_path(&self, image_path: String) -> AppResult<String> {
        let project_path = Self::get_project_path()?;
        let full_path = PathBuf::from(&project_path).join(&image_path);
        Ok(full_path.to_string_lossy().to_string())
    }
}

impl Default for ImageService {
    fn default() -> Self {
        Self::new()
    }
}
